controllers/topology/reconcile_state.go

```diff
@@ -62,17 +66,158 @@ func (r *ClusterReconciler) reconcileCluster(ctx context.Context, current, desir
+	}
+
+	if _, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		desired: md.bootstrapTemplate,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s", md.object.GroupVersionKind(), md.object.Name)
+	}
+
+	log.Info("creating", md.object.GroupVersionKind(), md.object.GetName())
+	if err := r.Client.Create(ctx, md.object.DeepCopy()); err != nil {
+		return errors.Wrapf(err, "failed to create the %s %s", md.object.GroupVersionKind(), md.object.Name)
+	}
+	return nil
+}
+
+// updateMachineDeployment updates a MachineDeployment. Also rotates the corresponding Templates if necessary.
+func (r *ClusterReconciler) updateMachineDeployment(ctx context.Context, clusterName string, currentMD, desiredMD machineDeploymentTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	cleanupOldInfrastructureTemplate, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		ref:     &desiredMD.object.Spec.Template.Spec.InfrastructureRef,
+		current: currentMD.infrastructureMachineTemplate,
+		desired: desiredMD.infrastructureMachineTemplate,
+		templateNamer: func() string {
+			// TODO: FIXME: Move the name generation into a shared func after #5070 has been merged
+			return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-infra-", clusterName, desiredMD.object.Name))
+		},
+		compatibilityChecker: checkReferencedObjectsAreCompatible,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+
+	cleanupOldBootstrapTemplate, err := r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+		ref:     desiredMD.object.Spec.Template.Spec.Bootstrap.ConfigRef,
+		current: currentMD.bootstrapTemplate,
+		desired: desiredMD.bootstrapTemplate,
+		templateNamer: func() string {
+			// TODO: FIXME: Move the name generation into a shared func after #5070 has been merged
+			return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-bootstrap-", clusterName, desiredMD.object.Name))
+		},
+		compatibilityChecker: checkReferencedObjectsAreInTheSameNamespace,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update the %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+
+	// Check differences between current and desired MachineDeployment, and eventually patch the current object.
+	// TODO @fabrizio Do we also need the DeepCopy here?
+	patchHelper, err := newMergePatchHelper(currentMD.object, desiredMD.object, r.Client)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create patch helper for %s %s", currentMD.object.GroupVersionKind(), currentMD.object.Name)
+	}
+	if patchHelper.HasChanges() {
+		log.Info("updating", currentMD.object.GroupVersionKind(), currentMD.object.GetName())
+		// TODO @fabrizio I see a few {"metadata":{"creationTimestamp":null}} patches in unit tests, I think it will also be there
+		// in "reality", any follow-up tasks?
```

Comment:
@fabrizio I see a few `{"metadata":{"creationTimestamp":null}}` patches in unit tests, I think it will also be there in "reality", any follow-up tasks?

Reply:
This should not happen.
Could you kindly check if by chance you are setting creationTimestamp during in the desired objects (in your case with the objects builders) or if you are altering the desired objects via fakeClient calls (and eventually re-use them in different test cases)?.

Reply:
I tested it in isolation and the problem is definitely there (but I think we need a separate issue).
Variable content during `newMergePatchHelper`
```
originalJSON:
{"kind":"MachineDeployment","apiVersion":"cluster.x-k8s.io/v1alpha4","metadata":{"name":"md-8-delete","namespace":"default","resourceVersion":"999","creationTimestamp":null},"spec":{"clusterName":"","selector":{},"template":{"metadata":{},"spec":{"clusterName":"","bootstrap":{"configRef":{"kind":"FakeBoostrapTemplate","namespace":"default","name":"bootstrap-config-8-delete","apiVersion":"bootstrap.cluster.x-k8s.io/v1alpha4"}},"infrastructureRef":{"kind":"FakeInfrastructureMachineTemplate","namespace":"default","name":"infrastructure-machine-8-delete","apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha4"}}}},"status":{}}

modifiedJSON:
{"kind":"MachineDeployment","apiVersion":"cluster.x-k8s.io/v1alpha4","metadata":{"name":"md-8-update","namespace":"default","creationTimestamp":null},"spec":{"clusterName":"","selector":{},"template":{"metadata":{},"spec":{"clusterName":"","bootstrap":{"configRef":{"kind":"FakeBoostrapTemplate","namespace":"default","name":"cluster-1-md-8-update-bootstrap-cpqw8","apiVersion":"bootstrap.cluster.x-k8s.io/v1alpha4"}},"infrastructureRef":{"kind":"FakeInfrastructureMachineTemplate","namespace":"default","name":"cluster-1-md-8-update-infra-h9ctm","apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha4"}}}},"status":{}}

originalWithModifiedJSON:
{"apiVersion":"cluster.x-k8s.io/v1alpha4","kind":"MachineDeployment","metadata":{"name":"md-8-update","namespace":"default","resourceVersion":"999"},"spec":{"clusterName":"","selector":{},"template":{"metadata":{},"spec":{"bootstrap":{"configRef":{"apiVersion":"bootstrap.cluster.x-k8s.io/v1alpha4","kind":"FakeBoostrapTemplate","name":"cluster-1-md-8-update-bootstrap-cpqw8","namespace":"default"}},"clusterName":"","infrastructureRef":{"apiVersion":"infrastructure.cluster.x-k8s.io/v1alpha4","kind":"FakeInfrastructureMachineTemplate","name":"cluster-1-md-8-update-infra-h9ctm","namespace":"default"}}}},"status":{}}

data:
{"metadata":{"creationTimestamp":null,"name":"md-8-update"},"spec":{"template":{"spec":{"bootstrap":{"configRef":{"name":"cluster-1-md-8-update-bootstrap-cpqw8"}},"infrastructureRef":{"name":"cluster-1-md-8-update-infra-h9ctm"}}}}}
```

So my interpretation:
* creationTimestamp is nil in the marshalled JSON (both for current and desired)
* originalWithModifiedJSON creationTimestamp is dropped there
* as a result the merge patch from originalJSON to originalWithModifiedJSON contains the creationTimestamp

Note: CreationTimestamp is not a pointer `CreationTimestamp Time `json:"creationTimestamp,omitempty" protobuf:"bytes,8,opt,name=creationTimestamp"` (also not in the debugger), it's just marshalled as nil.

Reply:
We debugged it locally. This only occurs when creationTimestamp is not set. But we're only using the patchHelper when we want to patch an already existing object => tl;dr not a problem we have in reality